### PR TITLE
[#7158] Determine RTLD flags from DT_FLAGS_1 for loading plugins

### PR DESCRIPTION
**Pull request by SwooshyCueb**
Addresses #7158 for main.
Prerequisite for irods/irods_rule_engine_plugin_python#157

### CODE REVIEW

**Comment by SwooshyCueb** [comment]
Still running tests, but they're far enough along that I think it's safe to mark this PR as ready for review.

**Comment by SwooshyCueb** [comment]
I think we should document this functionality and how to use it, but I'm not sure where

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (trel) [inline lib/core/include/irods/irods_load_plugin.hpp:178]: confirming this is `warn` on purpose.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (trel) [inline lib/core/include/irods/irods_load_plugin.hpp:191]: same, warn vs error.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:178]: I used `warn` because the situation is less severe than than the cases we use `error` for above, but either one makes sense here to me.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:191]: `warn` is intentional here. The absence of a section table doesn't necessarily mean there's something wrong, just that we haven't taught the server how to find `DT_FLAGS_1` without one.

Comment (trel) [reply lib/core/include/irods/irods_load_plugin.hpp:191]: great.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (trel) [reply lib/core/include/irods/irods_load_plugin.hpp:178]: got it - up to you.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:108]: Let's stick with `///` since that appears to have become the convention.

Also, missing `\param[in] _plugin_path description ...`, `\throws`, and `\returns`.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (alanking) [inline lib/core/include/irods/irods_load_plugin.hpp:124]: Our regular/default log messages use the key "log_message". Should we use the same key for consistency?

I realize that by now this may not be consistent everywhere, but figured I'd ask.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:110]: Consider dropping the use of `noexcept`.

The logger can throw exceptions.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:188]: If the section table isn't available, will that affect the PREP in regard to the DT_FLAGS_1 solution you've built?

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:204]: For non-const variables, let's invert the operands on each side of `==` to guard against future modifications.
```c++
0 == e_shnum || ...
```

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:279]: So `&e_strtbl_vec[sh_name]` doesn't work?

Is the pointer returned null-terminated?

**Review by alanking — COMMENTED** [review]
Seems fine overall in that nothing really jumped out at me. I might need to go read the ELF spec :) Just saw one thing which may or may not matter.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:124]: Good catch. I used `log_message` in PREP so I'll use that here too.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:339]: Just curious. Why `!= 0U` instead of `== DF_1_NOW`?

Are you interested in a subset of the bits being set?

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:370]: Let's add a trailing comment containing the name of this function here.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [inline lib/core/include/irods/irods_load_plugin.hpp:110]: Consider adding comments to the most important parts of this function. It's just nice not having to find documentation for all the scary data structures to understand what is happening here.

**Review by korydraughn — COMMENTED** [review]
Very nice implementation.

Once approved, please capture the TODOs in issues and link to the relevant parts of the implementation.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:188]: Yes. Without a section table, we can't currently get to the dynamic entry with `DT_FLAGS_1`, so the plugin would be loaded with the default flags.

I don't see this coming up any time soon. We presently only support x86_64 on Linux, so there will always be a section table. Nonetheless, I'll be opening an issue covering the TODOs once everything else passes review.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:279]: > So `&e_strtbl_vec[sh_name]` doesn't work?

My understanding is that depends on the implementation of `std::vector`, but I could be mistaken.

> Is the pointer returned null-terminated?

Yes.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:339]: Yes, if any other flags are also set, `== DF_1_NOW` would return false, even if `DF_1_NOW` is set.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:204]: You're gonna have to unwrap that one for me. How does switching the operands guard against future modifications?

Comment (trel) [reply lib/core/include/irods/irods_load_plugin.hpp:204]: just protects against fat-fingering the `==` to an assignment...

https://en.wikipedia.org/wiki/Yoda_conditions

Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:204]: That makes sense now that I think about it. I'll make the changes.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp]: To my understanding, the only potential source of exceptions from this function would be the logging in the exception handlers, so I'm not sure what to put for `\throws`.

Comment (korydraughn) [reply lib/core/include/irods/irods_load_plugin.hpp]: The logger uses nlohmann::json, fmtlib, etc.

I think std::exception is good enough. At the very least, anyone calling this needs to catch std::exception to be safe.

If it helps, I've started leaving out **why** a particular exception may be thrown. I find that the message can quickly become incorrect over time and therefore it's safer to just let the dev know which exceptions can be thrown. They can interrogate the `.what()` for the reason.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [reply lib/core/include/irods/irods_load_plugin.hpp:194]: Ok. That works for me.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [reply lib/core/include/irods/irods_load_plugin.hpp:357]: Unless I'm misunderstanding, the bitwise-and with `DF_1_NOW` will only return the bits matching `DF_1_NOW`.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (korydraughn) [reply lib/core/include/irods/irods_load_plugin.hpp]: Yes. Please mark as resolved.

File: lib/core/include/irods/irods_load_plugin.hpp
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_load_plugin.hpp:357]: oh, I misunderstood your question.

clang-tidy suggested `!= 0U`, so that's what I did. But `== DF_1_NOW` seems more correct. I'll make the changes.

**Review by korydraughn — COMMENTED** [review]
I think we're in a good spot.

Please squash without the pounds and we'll do one last look over it before approving.

**Comment by SwooshyCueb** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
Please add the pound.

**Review by alanking — APPROVED** [review]

**Comment by SwooshyCueb** [comment]
Looks like there may be a problem with Centos 7, so I'm going to hold off until I can figure it out.

**Comment by SwooshyCueb** [comment]
The problem was a test I added relying on a Python module that didn't exist yet. I've removed the test in question. https://github.com/irods/irods_rule_engine_plugin_python/pull/157#issuecomment-1623880239

\#'d

**Review by alanking — APPROVED** [review]

**Pull request merged**
